### DISCUSSION TITLE
[Thales MAV5.0-5.1 Java Card OS] Failed to Open Secure Channel, Card Cryptogram Invalid!

### DISCUSSION THREAD

**Discussion by traxformania** [martinpaljak/GlobalPlatformPro#336] [Q&A]
...

**Comment by martinpaljak** [comment]
Interesting card - can you send samples?

As the output tells:

https://github.com/martinpaljak/GlobalPlatformPro/wiki/Keys#diversification

I think the documentation on keys needs to be updated to clarify derivation. Quite often cards use proprietary derivation schemes (so don't use kdf3: unless you are absolutely sure that is the derivation scheme) which work with their software, but is hard/difficult to derive without additional knowledge from vendor documentation. Please consult with the card vendor and documentation to know which derivation scheme is needed. If the scheme is publicly documented and does not depend "out of band" input (INITIALIZE UPDATE has the derivation data must include it) it can be added to GPPro

**Reply by martinpaljak** [reply]
What you copied here is _session keys_ (GP defined) not _card keys_. How do you know your card keys are correct? How do you derive them? Please provide _that_ documentation.
